src/ng/directive/as.js

```diff
@@ -0,0 +1,90 @@
+'use strict';
+
+/**
+ * @ngdoc directive
+ * @name as
+ * @restrict A
+ *
+ * @description
+ * Publishes a component controller or a DOM element into the current scope.
+ *
+ * @element ANY
+ * @param {expression} as {@link guide/expression Expression} to assign the controller.
+ *
+ * This directive allows to access to the controller of components in your
+ * current template or from current controller.
+ *
+ * @example
+ * ### Access to component controller
+ * This example demonstrates basic publishing of the component controller into the scope.
+ * <example name="asDirectiveComponentExample" module="asComponentExample">
+ *   <file name="index.html">
+ *     <script>
+ *       angular.module('asComponentExample', [])
+ *        .component('toggle', {
+ *           controller: function() {
+ *             var opened = false;
+ *             this.isOpen = function() { return opened; };
+ *             this.toggle = function() { opened = !opened; };
+ *           }
+ *         });
+ *     </script>
+ *     <toggle as="myToggle"></toggle>
+ *     <button ng-click="myToggle.toggle()">Toggle</button>
+ *     <div ng-show="myToggle.isOpen()">You are using a children component to show it.</div>
+ *   </file>
+ *   <file name="protractor.js" type="protractor">
+ *      it('should publish the toggle into the scope', function() {
+ *        var toggle = element(by.buttonText('Toggle'));
+ *        expect(toggle.evaluate('myToggle.isOpen()')).toEqual(false);
+ *        toggle.click();
+ *        expect(toggle.evaluate('myToggle.isOpen()')).toEqual(true);
+ *      });
+ *   </file>
+ * </example>
+ *
+ * This directive also allows to bind elements to the scope.
+ *
+ * ```js
+ *   var myMod = angular.module(...);
+ *   myMod.component('myPlayer', {
+ *     template: '' +
+ *       '<audio as="$ctrl.audio" src="foo.ogg"></audio>' +
+ *       '<button ng-click="$ctrl.play()">Play</button>' +
+ *       '',
+ *     controller: function() {
+ *       this.play = function() {
+ *         this.audio.play();
+ *       };
+ *     }
+ *   });
+ * ```
+ *
+ * Note that angular does not allows to access DOM elements directly
+ * from templates, you need a controller to manipulate them.
+ *
+ */
+var componentAsDirective = ['$parse', function($parse) {
+  return {
+    restrict: 'A',
+    compile: function(tElement, tAttrs) {
+      // gets the theoretical controller name, converts <some-thing> into "someThing")
+      var controllerName = tElement[0].tagName.toLowerCase()
+        .replace(/[^\w]\w/g, function(a) { return a.slice(1).toUpperCase(); });
```

Comment:
Ok, I'll add the test and fix it.

El dv, 19 febr 2016 a les 14:08 Georgios Kalpakas <notifications@github.com>
va escriure:

> In src/ng/directive/as.js
> <https://github.com/angular/angular.js/pull/14080#discussion_r53458157>:
>
> > + *       };
> > + *     }
> > + *   });
> > + * ```
> > + *
> > + * Note that angular does not allows to access DOM elements directly
> > + * from templates, you need a controller to manipulate them.
> > + *
> > + */
> > +var componentAsDirective = ['$parse', function($parse) {
> > +  return {
> > +    restrict: 'A',
> > +    compile: function(tElement, tAttrs) {
> > +      // gets the theoretical controller name, converts <some-thing> into "someThing")
> > +      var controllerName = tElement[0].tagName.toLowerCase()
> > +        .replace(/[^\w]\w/g, function(a) { return a.slice(1).toUpperCase(); });
>
> This is not taking all possible cases into account (e.g. whe using
> data-/x- prefixes).
> You can use tAttrs.$normalize() instead.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/angular/angular.js/pull/14080/files#r53458157>.
>